File: skills/src/csharp/experimental/newsskill/Responses/Main/MainResponses.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.TemplateManager;
using Microsoft.Bot.Schema;

namespace NewsSkill.Responses.Main
{
    public class MainResponses : TemplateManager
    {
        // Constants
        public const string Cancelled = "cancelled";
        public const string Completed = "completed";
        public const string Confused = "confused";
        public const string Greeting = "greeting";
        public const string Help = "help";
        public const string Intro = "intro";

        private static LanguageTemplateDictionary _responseTemplates = new LanguageTemplateDictionary
        {
            ["default"] = new TemplateIdMap
            {
                { Cancelled, (context, data) => MainStrings.CANCELLED },
                { Completed, (context, data) => MainStrings.COMPLETED },
                { Confused, (context, data) => MainStrings.CONFUSED },
                { Greeting, (context, data) => MainStrings.GREETING },
                { Help, (context, data) => SendHelpCard(context, data) },
                { Intro, (context, data) => SendIntroCard(context, data) },
            },
            ["en"] = new TemplateIdMap { },
            ["fr"] = new TemplateIdMap { },
        };

        public MainResponses()
        {
            Register(new DictionaryRenderer(_responseTemplates));
        }

        public static IMessageActivity SendIntroCard(ITurnContext turnContext, dynamic data)
        {
            var response = turnContext.Activity.CreateReply();

            response.Attachments = new List<Attachment>()
            {
                new HeroCard()
                {
                    Title = MainStrings.INTRO_TITLE,
                    Text = MainStrings.INTRO_TEXT
                }.ToAttachment()
            };

            return response;
        }

        public static IMessageActivity SendHelpCard(ITurnContext turnContext, dynamic data)
        {
            var response = turnContext.Activity.CreateReply();
            response.Attachments = new List<Attachment>
            {
                new HeroCard()
                {
                    Title = MainStrings.HELP_TITLE,
                    Text = MainStrings.HELP_TEXT,
                    Buttons = new List<CardAction>()
                {
                    new CardAction(type: ActionTypes.ImBack, title: "Test", value: "Hello"),
                    new CardAction(type: ActionTypes.OpenUrl, title: "Learn More", value: "https://docs.microsoft.com/en-us/azure/bot-service/?view=azure-bot-service-4.0"),
                },
                }.ToAttachment()
            };

            return response;
        }
    }
}

File: templates/Virtual-Assistant-Template/typescript/samples/sample-assistant/src/responses/mainResponses.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import {
    Activity,
    Attachment,
    CardFactory,
    InputHints,
    MessageFactory,
    TurnContext } from 'botbuilder';
import { ActionTypes } from 'botframework-schema';
import { readFileSync } from 'fs';
import i18next from 'i18next';
import { join } from 'path';
import {
    DictionaryRenderer,
    LanguageTemplateDictionary,
    TemplateFunction } from '../templateManager/dictionaryRenderer';
import { TemplateManager } from '../templateManager/templateManager';

export class MainResponses extends TemplateManager {

    // Declare here the type of properties and the prompts
    public static responseIds: {
        cancelled: string;
        completed: string;
        confused: string;
        greeting: string;
        help: string;
        newUserGreeting: string;
        returningUserGreeting: string;
    } = {
        cancelled: 'cancelled',
        completed: 'completed',
        confused: 'confused',
        greeting: 'greeting',
        help: 'help',
        newUserGreeting: 'newUser',
        returningUserGreeting: 'returningUser'
    };

    // Declare the responses map prompts
    private static readonly responseTemplates: LanguageTemplateDictionary = new Map([
        ['default', new Map([
            [MainResponses.responseIds.cancelled, MainResponses.fromResources('main.cancelled')],
            [MainResponses.responseIds.completed, MainResponses.fromResources('main.completed')],
            [MainResponses.responseIds.confused, MainResponses.fromResources('main.confused')],
            [MainResponses.responseIds.greeting, MainResponses.fromResources('main.greeting')],
            [MainResponses.responseIds.help, MainResponses.buildHelpCard],
            [MainResponses.responseIds.newUserGreeting, MainResponses.buildNewUserGreetingCard],
            [MainResponses.responseIds.returningUserGreeting, MainResponses.buildReturningUserGreetingCard]
        ])]
    ]);

    // Initialize the responses class properties
    public constructor() {
        super();
        this.register(new DictionaryRenderer(MainResponses.responseTemplates));
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    public static async buildNewUserGreetingCard(turnContext: TurnContext, data: any): Promise<any> {
        const introFileName: string = i18next.t('main.introGreetingFile');
        const introPath: string = join(__dirname, '..', 'content', introFileName);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const introCard: any = JSON.parse(readFileSync(introPath, 'UTF8'));
        const attachment: Attachment = CardFactory.adaptiveCard(introCard);
        const response: Partial<Activity> = MessageFactory.attachment(attachment, '', attachment.content.speak, InputHints.AcceptingInput);

        response.suggestedActions = {
            actions: [
                {
                    title: i18next.t('main.helpBtnText1'),
                    type: ActionTypes.ImBack,
                    value: i18next.t('main.helpBtnValue1')
                },
                {
                    title: i18next.t('main.helpBtnText2'),
                    type: ActionTypes.ImBack,
                    value: i18next.t('main.helpBtnValue2')
                },
                {
                    title: i18next.t('main.helpBtnText3'),
                    type: ActionTypes.OpenUrl,
                    value: i18next.t('main.helpBtnValue3')
                }
            ],
            to: []
        };

        return Promise.resolve(response);
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    public static async buildReturningUserGreetingCard(turnContext: TurnContext, data: any): Promise<any> {
        const introFileName: string = i18next.t('main.introReturningFile');
        const introPath: string = join(__dirname, '..', 'content', introFileName);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const introCard: any = JSON.parse(readFileSync(introPath, 'UTF8'));
        const attachment: Attachment = CardFactory.adaptiveCard(introCard);
        const response: Partial<Activity> = MessageFactory.attachment(attachment, '', attachment.content.speak, InputHints.AcceptingInput);

        response.suggestedActions = {
            actions: [
                {
                    title: i18next.t('main.helpBtnText1'),
                    type: ActionTypes.ImBack,
                    value: i18next.t('main.helpBtnValue1')
                },
                {
                    title: i18next.t('main.helpBtnText2'),
                    type: ActionTypes.ImBack,
                    value: i18next.t('main.helpBtnValue2')
                },
                {
                    title: i18next.t('main.helpBtnText3'),
                    type: ActionTypes.OpenUrl,
                    value: i18next.t('main.helpBtnValue3')
                }
            ],
            to: []
        };

        return Promise.resolve(response);
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    public static async buildHelpCard(turnContext: TurnContext, data: any): Promise<any> {
        const title: string = i18next.t('main.helpTitle');
        const text: string = i18next.t('main.helpText');
        const attachment: Attachment = CardFactory.heroCard(title, text);
        const response: Partial<Activity> = MessageFactory.attachment(attachment, text, InputHints.AcceptingInput);

        response.suggestedActions = {
            actions: [
                {
                    title: i18next.t('main.helpBtnText1'),
                    type: ActionTypes.ImBack,
                    value: i18next.t('main.helpBtnValue1')
                },
                {
                    title: i18next.t('main.helpBtnText2'),
                    type: ActionTypes.ImBack,
                    value: i18next.t('main.helpBtnValue2')
                },
                {
                    title: i18next.t('main.helpBtnText3'),
                    type: ActionTypes.OpenUrl,
                    value: i18next.t('main.helpBtnValue3')
                }
            ],
            to: []
        };

        return Promise.resolve(response);
    }

    private static fromResources(name: string): TemplateFunction {
        return (): Promise<string> => Promise.resolve(i18next.t(name));
    }
}
